lib/web_ui/lib/src/engine/text/measurement.dart

```diff
@@ -763,39 +766,69 @@ class LinesCalculator {
     }
 
     if (isHardBreak) {
-      _addLineBreak(lineEnd: chunkEnd);
+      _addLineBreak(lineEnd: chunkEnd, isHardBreak: true);
     }
     _chunkStart = chunkEnd;
   }
 
-  void _addLineBreak({@required int lineEnd}) {
-    final int indexWithoutNewlines = _excludeTrailing(
+  void _addLineBreak({
+    @required int lineEnd,
+    @required bool isHardBreak,
+  }) {
+    final int endWithoutNewlines = _excludeTrailing(
       _text,
       _lineStart,
       lineEnd,
       _newlinePredicate,
     );
-    lines.add(_text.substring(_lineStart, indexWithoutNewlines));
+    final int endWithoutSpace = _excludeTrailing(
+      _text,
+      _lineStart,
+      lineEnd,
+      _whitespacePredicate,
+    );
+    final int lineNumber = lines.length;
+    final EngineLineMetrics metrics = EngineLineMetrics.withText(
+      _text.substring(_lineStart, endWithoutNewlines),
+      hardBreak: isHardBreak,
+      width: measureSubstring(_lineStart, endWithoutSpace),
+      lineNumber: lineNumber,
+    );
+    lines.add(metrics);
     _lineStart = lineEnd;
     if (lines.length == _style.maxLines) {
       _reachedMaxLines = true;
     }
   }
 
+  /// Measures the width of a substring of [_text] starting from the index
+  /// [start] (inclusive) to [end] (exclusive).
+  ///
+  /// This method uses [_text], [_style] and [_canvasContext] to perform the
+  /// measurement.
+  double measureSubstring(int start, int end) {
+    return _measureSubstring(_canvasContext, _style, _text, start, end);
+  }
+
   /// In a continuous block of text, finds the point where text can be broken to
   /// fit in the given constraint [maxWidth].
   ///
   /// This always returns at least one character even if there isn't enough
   /// space for it.
-  int _forceBreak(double maxWidth, String text, int start, int end) {
-    assert(0 <= start && start < end && end <= text.length);
+  int forceBreakSubstring({
+    @required double maxWidth,
+    @required int start,
+    @required int end,
+  }) {
+    assert(0 <= start && start < end && end <= _text.length);
```

Comment:
nit: when this assert fails it will say that `0 <= start && start < end && end <= _text.length` evaluated to false. I think breaking it up into separate asserts - `assert(0 <= start); assert(start < end); assert(end <= _text.length)` - would provide a more actionable error message.

Reply:
Done. Also applied the same change to few other asserts in this file.